daffodil-runtime1/src/main/scala/org/apache/daffodil/processors/parsers/SeparatedSequenceParsers.scala

```diff
@@ -0,0 +1,549 @@
+        //
+        // Not allowed to have any empty at all
+        //
+        failedZeroLengthWithAnyEmpty(pstate, priorState)
+      } else {
+        ParseAttemptStatus.Success_ZeroLength // ZL, successful parse of required element
+      }
+    } else {
+      ParseAttemptStatus.Success_NotZeroLength // not ZL, successful parse of required element
+    }
+  }
+
+  private def processSuccessfulOptionalVariableOccursChildParseResults(trd: TermRuntimeData,
+    parser: SeparatedChildParser,
+    zl: Boolean,
+    pstate: PState,
+    priorState: PState.Mark,
+    startState: PState.Mark,
+    ais: GoArrayIndexStatus): ParseAttemptStatus = {
+    if (zl) {
+      val shouldRemoveZLElement =
+        trd match {
+          case erd: ElementRuntimeData =>
+            parser.shouldRemoveZLStringHexBinaryValue(ais, erd)
+          case mgd: ModelGroupRuntimeData =>
+            false
+        }
+      if (shouldRemoveZLElement) {
+        // it's an optional element, type is string/hexBinary and length is zero
+        // so we don't want to add it to the infoset
+        //
+        // However, we do want to keep trying to parse more, as they could be trailing separators.
+        // that are to be tolerated.
+        //
+        // So we don't backtrack here. We just remove the accumulated element.
+        //
+        pstate.reset(priorState) // no need discard priorState, that is implicitly discarded by resetting the startState
+        pstate.discard(startState) // finished with array, so cleanup
+        ParseAttemptStatus.Success_EndOfArray // success, but we're at end of array. not zero length (in this case with success at prior element)
+      } else {
+        //
+        // optional, zero length, successful parse, not a string/hexBinary
+        //
+        // This can happen if you parse a choice group that
+        // has all its branches entirely optional
+        //
+        if (ssp eq SeparatorSuppressionPolicy.AnyEmpty) {
+          //
+          // Not allowed to have any empty at all
+          //
+          failedZeroLengthWithAnyEmpty(pstate, priorState)
+        } else {
+          ParseAttemptStatus.Success_ZeroLength // ZL, successful parse of optional 'thing'
+        }
+      }
+    } else {
+      ParseAttemptStatus.Success_NotZeroLength // not ZL, successful optional element parse
+    }
+  }
+
+  protected def parse(pstate: PState): Unit = {
+
+    var scpIndex = 0
+    pstate.mpstate.groupIndexStack.push(1L) // one-based indexing
+
+    val limit = children.length
+
+    var wasLastChildZeroLength = false
+
+    while ((scpIndex < limit) && (pstate.processorStatus eq Success)) {
+      val child = children(scpIndex)
+      child match {
+        case parser: RepSeparatedChildParser => {
+
+          val loopState = parser.loopState(pstate)
+
+          // push new array context for array/optional
+          parser.startArray(pstate)
+
+          val maybeStartState =
```

Comment:
Some thoughts on the mark leak since this is notoriously hard to track down.

I think this might be leaked. If a parse1 call throws an exception in parseOne, the processFailedChildParseResults method won't get called, which is supposed to discard/reset marks. The tryParseDetectMarkLeaks method will clean up the priorState mark leaked, but does not handle the maybeStartState. So I *think* you just need to discard the maybeStartState when priorState is discarded.

Anothe potential leak, in processSuccessfulChildParseResults, it looks like the state isn't reset/discarded if isFixedOccurs. I think some other code paths related to processSuccessful might be missing discards as well.

Reply:
All refactored away. I still have a few tests failing, but so far nothing appears to be leaking - in parsing anyway.